urld829b80034_wjch111_PTSP2013Wang_tgz-pJ8-controllers_momcts_momctsCore_TestPlannerJ8/src/controllers/momcts/utils/SetOperation.java

```diff
@@ -0,0 +1,673 @@
+			BigDecimal b = new BigDecimal(bs.get(i).toString());
+			multiplication.add(java.lang.Double.valueOf(a.multiply(b).toString()));
+		}
+		return multiplication;
+	}
+	
+	public static Vector<Double> muls(Vector<Double> as, double b){
+		Vector<Double> multiplication = new Vector<Double>();
+		for(int i=0; i< as.size();i++){
+			multiplication.add(as.get(i)*b);
+		}
+		return multiplication;
+	}
+	
+	public static Vector<Double> divs(Vector<Double> as, double b){
+		Vector<Double> multiplication = new Vector<Double>();
+		for(int i=0; i< as.size();i++){
+			multiplication.add(as.get(i)/b);
+		}
+		return multiplication;
+	}
+	
+	//---------end of Vector arithmatic Operators-----------
+	
+	//---------Map arithmatic Operators-----------
+	public static <T> HashMap<T, Double> muls(Map<T, Double> mp, double b){
+		HashMap<T, Double> mulMp = new HashMap<T, Double>();
+		for(T k : mulMp.keySet()){
+			mulMp.put(k, mp.get(k)*b);
+		}
+		return mulMp;
+	}
+
+	public static <T> HashMap<T, Double> divs(Map<T, Double> mp, double b){
+		HashMap<T, Double> divMp = new HashMap<T, Double>();
+		for(T k : mp.keySet()){
+			divMp.put(k, Presentation.ndigits(mp.get(k)/b,3));
+		}
+		return divMp;
+	}
+	//---------end of Map arithmatic Operators-----------
+	
+	//---------vector operations----------
+	public static <T> T pop(Vector<T> v, int elemInd){
+		T elem = v.get(elemInd);
+		v.remove(elemInd);
+		return elem;
+	}
+	
+	public static <T> T popFront(Vector<T> v){
+		return pop(v, 0);
+	}
+	
+	public static <T> T popEnd(Vector<T> v){
+		return pop(v, v.size()-1);
+	}
+	
+	public static <T> Vector<T> subSeq(Vector<T> v, int beginInd, int rangeLength){
+		Vector<T> subS = new Vector<T>();
+		int initPos = Math.max(0, beginInd);
+		int endPos = Math.min(v.size(), initPos+rangeLength);
+		for(int i= initPos; i< endPos; i++){
+			subS.add(v.get(i));
+		}
+		return subS;
+	}
+	
+	public static <T> Vector<T> subSeq(Vector<T> v, int beginInd){
+		return subSeq(v, beginInd, v.size());
+	}
+	
+	public static <T> Vector<T> firstNElements(Vector<T> v, int n){
+		return subSeq(v, 0, n);
+	}
+	
+	public static <T> boolean exists(Vector<T> vt, int pos){
+		if(pos >= 0 && pos < vt.size()){
+			return true;
+		} else {
+			return false;
+		}
+	}
+	
+	public static <T> T randomElement(Vector<T> v){
+		if(v.size()>0){
+			return v.get(MathOperation.rndInt(v.size()));	
+		} else {
+			return null;
+		}
+	}
+	
+	public static <T> Vector<T> joinSeq(Vector<T> v1, Vector<T> v2){
+		@SuppressWarnings("unchecked")
+		Vector<T> vjoin = (Vector<T>) v1.clone();
+		vjoin.addAll(v2);
+		return vjoin;
+	}
+	
+	/**
+	 * Join two sequences and limit their total length under maxseq, if surpassed, cut the head of the joint sequence
+	 * @param s
+	 * @param segment
+	 * @param maxseq
+	 * @return
+	 */
+	public static <T> Vector<T> joinSeq(Vector<T> s, Vector<T> segment, int maxseq){
+		@SuppressWarnings("unchecked")
+		Vector<T> seqOrg = (Vector<T>) s.clone();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;s&#39; (index: 0 | type: Vector). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;maxseq&#39;, &#39;segment&#39;. The same parameter type is validated here for example: urld829b80034_wjch111_PTSP2013Wang_tgz-pJ8-controllers_momcts_momctsCore_TestPlannerJ8/src/controllers/momcts/MOOTools.java:29. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).